lib/Sema/CSApply.cpp

```diff
@@ -9127,20 +9127,22 @@ static llvm::Optional<SyntacticElementTarget> applySolutionToForEachStmt(
   auto *parsedSequence = stmt->getParsedSequence();
   bool isAsync = stmt->getAwaitLoc().isValid();
 
-  // Simplify the various types.
-  forEachStmtInfo.sequenceType =
-      solution.simplifyType(forEachStmtInfo.sequenceType);
-  forEachStmtInfo.elementType =
-      solution.simplifyType(forEachStmtInfo.elementType);
-  forEachStmtInfo.initType =
-      solution.simplifyType(forEachStmtInfo.initType);
-
   auto &cs = solution.getConstraintSystem();
   auto *dc = target.getDeclContext();
 
-  // First, let's apply the solution to the sequence expression.
-  {
-    auto *makeIteratorVar = forEachStmtInfo.makeIteratorVar;
+  if (forEachStmtInfo.isa<SequenceIterationInfo>()) {
```

Comment:
I see that you turned CSGen/CSApply methods into giant if/else blocks - It would be better to use an early return pattern when possible because it leaves unrelated code unmodified (which is a lot easier to review because it leaves only functional changes), to archive that you'd need to encapsulate new logic into a method and call it with an early return.

Reply:
I have encapsulated the logic into a separate method. However, we can't take advantage of an early return pattern here since there's some shared functionality that has to be done in the end (e.g. constraints for the `where` expression)